### ISSUE TITLE
Joining a room with a problematic client causes the entire room to throw an error

### ISSUE THREAD

**Issue by christopher-abastar** [miguelgrinberg/Flask-SocketIO#243]
Hello,

I tried running the flask socket.io via gunicorn (no message queues or redis this time due to eventlet bug). We left it there for testing, but whenever a client joins a room named "Main" it disconnects and throws some invalid frame header (in chrome) or bad request when polling. However, joining a different room other than the room above seems to be ok.

I'm guessing one of the client registered in the server is having some issues, causing other connecting clients to that room having issues as well.

``` python
[2016-03-18 03:23:38 +0000] [24862] [ERROR] Error handling request /socket.io/?transport=websocket&sid=af3d602bb01744eaa94f802900aa8aac
Traceback (most recent call last):
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 335, in _get_socket
    s = self.sockets[sid]
KeyError: 'f27cbeb470fe4da8b38690d4f8db88e8'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vf/env/lib/python3.5/site-packages/gunicorn/workers/async.py", line 52, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/home/vf/env/lib/python3.5/site-packages/gunicorn/workers/async.py", line 112, in handle_request
    for item in respiter:
  File "/home/vf/env/lib/python3.5/site-packages/newrelic-2.58.2.45/newrelic/api/web_transaction.py", line 717, in __iter__
    for item in self.generator:
  File "/home/vf/env/lib/python3.5/site-packages/newrelic-2.58.2.45/newrelic/api/web_transaction.py", line 1093, in __call__
    self.start_response)
  File "/home/vf/env/lib/python3.5/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/flask_socketio/__init__.py", line 37, in __call__
    start_response)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 303, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 226, in handle_request
    environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/socket.py", line 79, in handle_get_request
    start_response)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/socket.py", line 118, in _upgrade_websocket
    return ws(environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/async_eventlet.py", line 13, in __call__
    return super(WebSocketWSGI, self).__call__(environ, start_response)
  File "/home/vf/env/lib/python3.5/site-packages/eventlet/websocket.py", line 127, in __call__
    self.handler(ws)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/socket.py", line 180, in _websocket_handler
    self.receive(pkt)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/socket.py", line 48, in receive
    self.server._trigger_event('message', self.sid, pkt.data)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 330, in _trigger_event
    return self.handlers[event](*args)
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 429, in _handle_eio_message
    self._handle_event(sid, pkt.namespace, pkt.id, pkt.data)
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 371, in _handle_event
    r = self._trigger_event(data[0], namespace, sid, *data[1:])
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 397, in _trigger_event
    return self.handlers[namespace][event](*args)
  File "/home/vf/env/lib/python3.5/site-packages/flask_socketio/__init__.py", line 190, in _handler
    ret = handler(*args)
  File "/usr/home/vf/releases/20160317093907/gameapp/app/socket/connect.py", line 16, in wrapped
    return f(*args, **kwargs)
  File "/usr/home/vf/releases/20160317093907/gameapp/app/socket/chat.py", line 14, in on_join
    send(socketio.user_id + ' has entered the room.', room=room)
  File "/home/vf/env/lib/python3.5/site-packages/flask_socketio/__init__.py", line 565, in send
    include_self=include_self, callback=callback)
  File "/home/vf/env/lib/python3.5/site-packages/flask_socketio/__init__.py", line 320, in send
    skip_sid=skip_sid, callback=callback)
  File "/home/vf/env/lib/python3.5/site-packages/flask_socketio/__init__.py", line 285, in emit
    callback=kwargs.get('callback'))
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 181, in emit
    self.manager.emit(event, data, namespace, room, skip_sid, callback)
  File "/home/vf/env/lib/python3.5/site-packages/socketio/base_manager.py", line 101, in emit
    self.server._emit_internal(sid, event, data, namespace, id)
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 322, in _emit_internal
    binary=binary))
  File "/home/vf/env/lib/python3.5/site-packages/socketio/server.py", line 333, in _send_packet
    self.eio.send(sid, encoded_packet, binary=False)
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 157, in send
    self._get_socket(sid).send(packet.Packet(packet.MESSAGE, data=data,
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 337, in _get_socket
    raise KeyError('Session not found')
KeyError: 'Session not found'
```

**Comment by miguelgrinberg** [comment]
This error occurs because somehow a disconnected user remained attached to the room, so the server tries to emit to it. It is possible that this problem is fixed already, I believe you are not running the latest versions of python-socketio and python-engineio. Please upgrade these packages and let me know if you still get failures.

**Comment by christopher-abastar** [comment]
Hi Miguel, 

Thanks for the quick reply.

I'm currently running the latest version for these packages:
```sh
python-engineio==0.9.0
python-socketio==1.1 
```

**Comment by miguelgrinberg** [comment]
Does this reproduce when you restart? If you can make it happen again, it might be useful for you to capture a logging session. If you add `logger=True, engineio_logger=True` to the `SocketIO` constructor logging should go to your terminal.

**Comment by christopher-abastar** [comment]
Hi Miguel,

I was able to grab some logs.

This user was disconnected a long time ago.  
```sh
30262de665a8490f8091a08223a5675f: Sending packet OPEN data {'sid': '30262de665a8490f8091a08223a5675f', 'pingTimeout': 60000, 'upgrades': ['websocket'], 'pingInterval': 25000}
30262de665a8490f8091a08223a5675f: Sending packet MESSAGE data 0
30262de665a8490f8091a08223a5675f: Received request to upgrade to websocket
30262de665a8490f8091a08223a5675f: Sending packet NOOP data None
30262de665a8490f8091a08223a5675f: Upgrade to websocket successful
30262de665a8490f8091a08223a5675f: Received packet MESSAGE data 2/ws,["authenticate","jwt"]
30262de665a8490f8091a08223a5675f: Received packet MESSAGE data 2/ws,["join",{"room":"Main"}]
30262de665a8490f8091a08223a5675f: Sending packet MESSAGE data 2/ws,["message","a3c5c600-b4eb-4b25-820a-54665a638b0d has entered the room."]
30262de665a8490f8091a08223a5675f: Client is gone, closing socket
```

Still, joining the the "Main" room for new clients will resolve to that user not being removed.
```
2016-03-18 07:35:06 +0000] [34637] [ERROR] Error handling request /socket.io/?EIO=3&transport=polling&t=LE8fZ6L&sid=6f657f6a7ee74312bbcf8ff480fac214
Traceback (most recent call last):
  File "/home/vf/env/lib/python3.5/site-packages/engineio/server.py", line 335, in _get_socket
    s = self.sockets[sid]
KeyError: '30262de665a8490f8091a08223a5675f'

During handling of the above exception, another exception occurred:
...
```

**Comment by miguelgrinberg** [comment]
Okay, thanks. I will investigate. The "Client is gone" message should have been accompanied with a "disconnect" callback, which should have removed the client from all rooms, but clearly that isn't working.

**Comment by mbach04** [comment]
@miguelgrinberg ::
Seeing the same problem on my end, but digging deeper it looks like this could be the answer? I've no idea how to apply it to the web server though.
http://stackoverflow.com/questions/22169036/websocket-permessage-deflate-in-chrome-with-no-context-takeover

**Comment by miguelgrinberg** [comment]
It looks like I forgot to update this issue, which I believe has been fixed in release 0.9.1 of python-engineio, commit https://github.com/miguelgrinberg/python-engineio/commit/95c9a55457e9cbd36597b14ad840e31abdb2030e.

If this is still a problem, please reopen.
